File: leetcode1726/solution.go
/*
Solution:

Time: O(n^2) | Space: O(n^2)

Runtime: 139 ms | 100.00%
Memory: 41.49 MB | 100.00%

- n: length of 'nums'
*/

func tupleSameProduct(nums []int) int {
	n := len(nums)
	m := make(map[int]int)

	res := 0

	for i := range n {
		for j := i + 1; j < n; j++ {
			prd := nums[i] * nums[j]

			m[prd] += 1
			res += (m[prd] - 1) * 8
		}
	}

	return res
}

File: leetcode1726/solution.rs
/*
Solution: 

Time: O(n^2) | Space: O(n^2)

Runtime: 57 ms | 100.00%
Memory: 11.26 MB | 50.00%

- n: length of 'nums'
*/

use std::collections::HashMap;

impl Solution {
    pub fn tuple_same_product(nums: Vec<i32>) -> i32 {
        let n: usize = nums.len();
        let mut map: HashMap<i32, i32> = HashMap::new();

        let mut res: i32 = 0;

        for i in 0..n {
            for j in (i + 1)..n {
                let prd: i32 = nums[i] * nums[j];

                map.entry(prd).and_modify(|cnt| {
                    res += *cnt * 8;
                    *cnt += 1;
                }).or_insert(1);
            }
        }

        res
    }
}